docs/olmv1_overview.md

```diff
@@ -0,0 +1,239 @@
+# OLMv1 Overview
+
+## What won't OLMv1 do that OLMv0 did?
+
+TL;DR: OLMv1 cannot feasibly support multi-tenancy or any feature that assumes multi-tenancy. All multi-tenancy features end up falling over because of the global API system of Kubernetes. While this short conclusion may be unsatisfying, the reasons are complex and intertwined.
+
+Nearly every engineer in the Operator Framework group contributed to design explorations and prototypes over an entire year. For each of these design explorations, there are complex webs of features and assumptions that are necessary to understand the context that ultimately led to a conclusion of infeasibility that led us to today’s conclusion.
+
+Here is a sampling of some of the ideas we explored:
+- [[WIP] OLM v1's approach to multi-tenancy](https://docs.google.com/document/d/1xTu7XadmqD61imJisjnP9A6k38_fiZQ8ThvZSDYszog/edit#heading=h.m19itc78n5rw)
+- [OLMv1 Multi-tenancy Brainstorming](https://docs.google.com/document/d/1ihFuJR9YS_GWW4_p3qjXu3WjvK0NIPIkt0qGixirQO8/edit#heading=h.vy9860qq1j01)
+
+### Watched namespaces cannot be configured in a first-class API
+
+OLMv1 will not have a first-class API for configuring the namespaces that a controller will watch.
+
+Kubernetes APIs are global. Kubernetes is designed with the assumption that a controller WILL reconcile an object no matter where it is in the cluster.
+
+However, Kubernetes does not assume that a controller will be successful when it reconciles an object.
+
+The Kubernetes design assumptions are:
+- CRDs and their controllers are trusted cluster extensions.
+- If an object for an API exists a controller WILL reconcile it, no matter where it is in the cluster.
+
+OLMv1 will make the same assumption that Kubernetes does and that users of Kubernetes APIs do. That is: If a user has RBAC to create an object in the cluster, they can expect that a controller exists that will reconcile that object. If this assumption does not hold, it will be considered a configuration issue, not an OLMv1 bug.
+
+This means that it is a best practice to implement and configure controllers to have cluster-wide permission to read and update the status of their primary APIs. It does not mean that a controller needs cluster-wide access to read/write secondary APIs. If a controller can update the status of its primary APIs, it can tell users when it lacks permission to act on secondary APIs.
+
+### Dependencies based on watched namespaces
+
+Since there will be no first-class support for configuration of watched namespaces, OLMv1 cannot resolve dependencies among bundles based on where controllers are watching.
```

Comment:
If we have the following scenario:

* Customer goes to install Operator A at the cluster scope (CRD + controller with cluster-wide RBAC)
* Operator A has a dependency on Operator B (A will be creating CRs from B and expects B to reconcile them)

I'm not sure why the absence of a watched namespaces concept prevents this dependency fulfillment.

OLMv1, given its limited RBAC, can likely only see that CRD B is not present. If it was present, OLMv1 couldn't see if controller B is available and reconciling.

OLMv1 however, if its RBAC was limited to a single namespace, could create a CR for B and see if controller B picks it up and sets some status field. This would give OLMv1 the information it needs about whether controller B is running and reconciling at the cluster scope (which as presented is the recommend -- possibly only? -- install mode).

Reply:
> Customer goes to install Operator A at the cluster scope (CRD + controller with cluster-wide RBAC)

OLMv1 will be completely unaware of the scoping configuration of Operator A. It doesn't know if Operator A is watching the entire cluster or is watching just a subset.

The fact that an opaque configuration applied to the bundle results in cluster-wide RBAC for the service account tied to a deployment is _maybe_ a good enough proxy for watch namespace. But you are correct that a dependency resolver would _also_ need awareness of the RBAC in use by any dependents, which is not available to OLM and may not be available to a user.

> OLMv1 however, if its RBAC was limited to a single namespace, could create a CR for B and see if controller B picks it up and sets some status field. This would give OLMv1 the information it needs about whether controller B is running and reconciling at the cluster scope (which as presented is the recommend -- possibly only? -- install mode).

This seems fragile and complex and lots could go wrong. Off the top of my head:
- OLM would have to evaluate the schema of each CRD and be able to produce a valid object
- This assumes that all objects have a status
- Operators may have admission webhooks that reject creates unless arbitrary conditions are met.
- Creating a CR may have major implications on the operations of a cluster, and would like incur costs.
- The fact that a CR for B can be created in a particular namespace provides no signal about whether a controller would reconcile B in another namespace.

The most likely dependency resolver implementation given the constraints is probably:
- Client-based
- Limited to cluster admins who can see RBAC and Extensions cluster-wide.
- Required to use RBAC as a proxy for watch namespaces, which may result in assumptions that a controller is watching a namespace, even if it is not (i.e. it has RBAC, but for whatever reason isn't actually watching there)